oml/functional/metrics.py

```diff
@@ -465,15 +489,213 @@ def calc_fnmr_at_fmr(pos_dist: torch.Tensor, neg_dist: torch.Tensor, fmr_vals: T
     return fnmr_at_fmr
 
 
+def calc_pcf(embeddings: torch.Tensor, explained_variance_ths: Tuple[float, ...]) -> List[torch.Tensor]:
+    """
+    Function estimates the Principal Components Fraction (PCF) of embeddings using Principal Component Analysis.
+    The number of components is estimated as a fraction of the embeddings dimension.
+
+    Args:
+        embeddings: Embeddings matrix with the shape of ``[n_embeddings, embeddings_dim]``.
+        explained_variance_ths: Values in range [0, 1]. Find the number of components such that the amount
+                                    of variance that needs to be explained is greater than the fraction specified
+                                    by ``explained_variance_ths``.
+    Returns:
+        List of linear dimensions as a fractions of the embeddings dimension.
+
+    Let :math:`X` be a set of :math:`d` dimensional embeddings.
+    Let :math:`\\lambda_1, \\ldots, \\lambda_d\\in\\mathbb{R}` be a set of eigenvalues
+    of the covariance matrix of :math:`X` sorted in descending order.
+    Then for a given value of desired explained variance :math:`r`,
+    the number of principal components that explaines :math:`r\\cdot 100\\%%` variance is the largest integer
+    :math:`n` such that
+
+    .. math::
+        \\frac{\\sum\\limits_{i = 1}^{n - 1}\\lambda_i}{\\sum\\limits_{i = 1}^{d}\\lambda_i} \\leq r
+
+    The function returns
+
+    .. math::
+        \\frac{n}{d}
+
+    See:
+
+        `Principal Components Analysis`_
+
+    .. _`Principal Components Analysis`:
+        https://en.wikipedia.org/wiki/Principal_component_analysis
+
+    Example:
+        In the example bellow there are 4 vectors of length 10, and only first 4 dimensions have non-zero values.
+        Its covariance matrix will have only 4 eigenvalues, that are greater than 0, i.e. there are only 4 principal
+        axis. So, in order to keep at least 50% of the information from the set, we need to keep just 2 dimensions
+        (i.e. 20% of the dimension of the original embeddings). And in order to keep all the information we need to
+        keep 50% of the dimension of the original embeddings.
+
+        >>> embeddings = torch.eye(4, 10, dtype=torch.float)
+        >>> calc_pcf(embeddings, explained_variance_ths=(0.5, 1))
+        tensor([0.2000, 0.5000])
+
+    """
+    # The code below mirrors code from scikit-learn repository:
+    # https://github.com/scikit-learn/scikit-learn/blob/f3f51f9b6/sklearn/decomposition/_pca.py#L491
+    _check_if_in_range(explained_variance_ths, 0, 1, "explained_variance_ths")
+    pca = PCA(embeddings)
+    ratio_cumsum = torch.cumsum(pca.explained_variance_ratio, dim=0)
+    n_components = (
+        torch.searchsorted(ratio_cumsum, torch.tensor(explained_variance_ths).to(embeddings), side="right") + 1
+    )
+    return n_components / embeddings.shape[1]
+
+
+class PCA:
+    """
+    Principal component analysis (PCA).
```

Comment:
Let's add our motivation for having this: Note, the code is almost the same as one from sklearn, but we had two reasons to have our own implementation. First, we need to work with Torch tensors instead of NumPy arrays. Second, we wanted to avoid one more external dependency.